src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -31,6 +31,47 @@ public static void Initialize()
             Dns.GetHostName();
         }
 
+        public static KeepAliveOption GetKeepAliveState(Socket socket)
+        {
+            bool enabled = (int)socket.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive) == 1;
+
+            // TcpKeepAliveRetryCount is 10 on Windows versions supported by .NET Core (Win 7 SP1 and later):
+            // https://msdn.microsoft.com/en-us/library/windows/desktop/ee470551.aspx
+            // It can be managed via *SocketOption starting from Windows 10 version 1703
+            Version version = Environment.OSVersion.Version;
+            bool isBelowWin10V1703 = version.Major < 10 || version.Major == 10 && version.Build < 15063;
+            int retryCount = isBelowWin10V1703 ? 10 : (int)socket.GetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount);
+
+            // TcpKeepAliveTime and TcpKeepAliveInterval can be managed via *SocketOption starting from Windows 10 version 1709
+            int time = (int)socket.GetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime);
+            int interval = (int)socket.GetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval);
+
+            return new KeepAliveOption(enabled, retryCount, time, interval);
+        }
+
+        public static void SetKeepAliveState(Socket socket, KeepAliveOption keepAliveOption)
+        {
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, keepAliveOption.Enabled);
+            if (!keepAliveOption.Enabled)
+                return;
+
+            // TcpKeepAliveRetryCount can be managed via *SocketOption starting from Windows 10 version 1703
+            // TcpKeepAliveTime and TcpKeepAliveInterval can be managed via *SocketOption starting from Windows 10 version 1709
+            Version version = Environment.OSVersion.Version;
+            bool isWin10V1703OrLater = version.Major > 10 || version.Major == 10 && version.Build >= 15063;
+            if (isWin10V1703OrLater)
+                socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveRetryCount, keepAliveOption.RetryCount);
```

Comment:
Why are we doing these version checks?  What happens if we're on an OS where it isn't supported?  I would prefer to make the call regardless, and if it fails, then either allow the error to propagate or eat it, but not try to guess at which version it'll work on.

Reply:
Probably this is not part of the approved API shape.
The SetSocketOption already throws if not supported.
My attempt was offering something that could be documented but used straight away instead of a user having to essentially write the code of this PR in his client code: as a user, not always expert of Socker internals, I would like to set keep-alive with a line of code